### DISCUSSION TITLE
RowDrop, get whole column in func

### DISCUSSION THREAD

**Discussion by fedorkobak** [pdpipe/pdpipe#94] [Q&A] [answered]
Nice work! I need some help. I need drop rows in data frame by condition depends on whole column, but RowDrop stage gives only single ints into function. For example:

```
import pandas as pd
import pdpipe as pdp
import numpy as np

df = pd.DataFrame({'a': ['a','b','c','d'], 'b':[5,6,7,1]})
def test_fun(x):

    return mean(x) < x

pdp.RowDrop( 
    [test_fun], 
    columns = pdp.cq.ColumnQualifier(lambda df: df.select_dtypes('number').columns))\
    .apply(df)
```
This code causes an error because `test_fun` gets separate integers, not whole column. For better understanding change `test_fun` like this:

```
def test_fun(x):
    print(type(x))
    return True
    #return mean(x) < x
```
It will show:
```
<class 'int'>
<class 'int'>
<class 'int'>
<class 'int'>
```

Is it any possibility for this case solving?

**Comment by fedorkobak** [comment]
Alternative solution may looks like
```
def test_fun(df):
    means = df.mean()
    res = np.zeros(df.shape[0], dtype = bool)
    for col in df:
        res = res | (df[col] < df[col].mean())
    
    return res

pdp.ColByFrameFunc("temp", test_fun).RowDrop({"temp":lambda x: x}).ColDrop('temp').apply(df)
```
But i used three stages instead of one, may there is better solution.

**Comment by shaypal5** [comment, marked as answer]
This is a very good question! :)

First of all, for a specific column the most concise way to write this would probably be as follows:
```python
pline2 = pdp.PdPipeline([
    pdp.ColByFrameFunc(
        column='b',
        func=lambda df: df['b'].apply(lambda x: 'DROP' if x < df['b'].mean() else x)
    ),
    pdp.ValDrop(['DROP']),
])
```

The first stage will mark all values to drop - and it operates at the dataframe level, and so can reference the entire column and calculate its mean. The second will then drop any row with the value `'DROP'` anywhere in it. Of course, if you want the column to keep its numeric dtype I would suggest you use a special numeric value.


If, however, you want this done for all numeric columns, the following way will achieve this with a bit more code - some of this I had to add to the package, and is now released as part of `v0.2.3`:

```python
pline = pdp.PdPipeline([
    pdp.ApplicationContextEnricher(
        numeric_means=lambda df: df.select_dtypes(include=np.number).mean().to_dict(),
    ),
    pdp.ApplyByCols(
        columns=pdp.cq.OfNumericDtypes(),
        func=lambda x, label, application_context: 'DROP' if x < application_context['numeric_means'][label] else x,
    ),
    pdp.ValDrop(['DROP']),
])
```

So what's new here is:

1. I've added the `ApplicationContextEnricher` stage which you can use to calculate dataframe-wide quantities, and any other data later stage might want to use, and make sure its calculated only once (and discarded after pipeline application).

2. I've now made sure that any `callable` supplied to `ApplyByCols` is injected with the column label, the application context and the fit context, if it so wishes (meaning, if it states the `label`, `application_context` or `fit_context` keyword arguments in its signature). So the first parameter of the callable is stil a cell-specific value (this is just how `pandas.DataFrame.apply` works), but the later parameters can give the callable a wider context, like which column am I operating one, or general dataframe- or pipeline-level quantities that were computed earlier during the pipeline's run and stored in its fit context (which is retained in the pipeline object for future transforms) or its application context (which, again, is created and discarded per transform).

I hope this helped you. Feel free to ask more! :)

Cheers,
Shy

**Reply by fedorkobak** [reply]
I'm guessing that in first example lambda function must looks like `lambda df: df['b'].apply(lambda x: 'DROP' if x < df['b'].mean() else x)` (df instead of tdf)?

**Reply by shaypal5** [reply]
Yeah, you're right! Fixed. I mean, the parameter name doesn't have to be `df`, but you do need to use the same name inside the lambda. :)

`lambda tdf: tdf['b'].apply(lambda x: 'DROP' if x < tdf['b'].mean() else x)` would also have worked.
